File: examples/ch03/Tree.hs
module Tree where

{-- snippet Tree --}
data Tree a = Node a (Tree a) (Tree a)
            | Empty
              deriving (Show)

{-- /snippet Tree --}

{-- snippet simpleTree --}
simpleTree = Node "parent" (Node "left child" Empty Empty)
                           (Node "right child" Empty Empty)
{-- /snippet simpleTree --}

File: examples/ch03/Tree.java
/** snippet Tree */
class Tree<A>
{
    A value;
    Tree<A> left;
    Tree<A> right;

    public Tree(A v, Tree<A> l, Tree<A> r)
    {
	value = v;
	left = l;
	right = r;
    }
}
/** /snippet Tree */

/** snippet Example */
class Example 
{
    static Tree<String> simpleTree()
    {
	return new Tree<String>(
            "parent",
	    new Tree<String>("left leaf", null, null),
	    new Tree<String>("right leaf", null, null));
    }
}
/** /snippet Example */
